simd/keccakf1600/f1600x4.go

```diff
@@ -1,22 +1,76 @@
-// f1600x4 implements a fast fourway interleaved Keccak-f[1600] permutation on
-// systems that support it.  Keccak-f[1600] is the permutation underlying
-// Keccak, SHA3 and SHAKE.
+// Package keccakf1600 provides a fourway Keccak-f[1600] permutation in parallel.
 //
-// This parallel f[1600] is useful in a few niche applications: for instance,
-// when computing many parallel SHAKE-128 hashes with same-length inputs such
-// as is done in Dilithium or XMSS.
+// Keccak-f[1600] is the permutation underlying several algorithms such as
+// Keccak, SHA3 and SHAKE. Running four permutations in parallel is useful in
+// some scenarios like in hash-based signatures.
 //
-// Currently only amd64 systems with AVX2 are supported.  This package does
-// not provide a generic implementation as it would be signficantly slower
-// than using non-interleaved f[1600].  Check keccakf1600.AvailableX4 to see
-// if the current system is supported.
+// Limitations
+//
+// Note that not all the architectures support SIMD instructions. This package
+// uses AVX2 instructions available in some AMD64 architectures.
+//
+// For those systems not supporting AVX2, the package still provides the
+// expected functionality by means of a generic and slow implementation.
+// The recommentation is to beforehand verify IsSIMDEnabled() to determine if
+// the current system supports the SIMD implementation.
 package keccakf1600
 
 import (
+	"unsafe"
+
 	"github.com/cloudflare/circl/internal/shake"
+	"golang.org/x/sys/cpu"
 )
 
-func genericF1600x4(a []uint64) {
+// StateX4 contains state for the fourway permutation including the four
+// interleaved [25]uint64 buffers. Call Initialize() before use to initialize
+// and get a pointer to the interleaved buffer.
+type StateX4 struct {
+	// Go guarantees a to be aligned on 8 bytes, whereas we need it to be
+	// aligned on 32 bytes for bet performance.  Thus we leave some headroom
+	// to be able to move the start of the state.
+
+	// 4 x 25 uint64s for the interleaved states and three uint64s headroom
+	// to fix allignment.
+	a [103]uint64
+
+	// Offset into a that is 32 byte aligned.
+	offset int
+}
+
+// IsSIMDEnabled returns true if the architecture supports SIMD implementation
+// provided in this package.
+func IsSIMDEnabled() bool { return cpu.X86.HasAVX2 }
```

Comment:
If we add `X8` support using AVX512, then this generic `IsSIMDEnabled` is not helpful.  `IsEnabledX4`?